### PR TITLE
arrow-cpp: 0.17.1 -> 1.0.0

### PR DESCRIPTION
**Pull request by tobim**
###### Motivation for this change
https://arrow.apache.org/blog/2020/07/24/1.0.0-release/

###### Things done
Fixed the static build with:
- Changing utf8proc to build with CMake.
- Adding changes of https://github.com/apache/arrow/pull/7818 and https://github.com/apache/arrow/pull/7842

###### Checks

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [x] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/libraries/arrow-cpp/default.nix [hunk 1]
```diff
@@ -1,26 +1,42 @@
-{ stdenv, lib, fetchurl, fetchFromGitHub, fetchpatch, fixDarwinDylibNames, autoconf, boost
-, brotli, cmake, flatbuffers, gflags, glog, gtest, lz4, perl
-, python3, rapidjson, snappy, thrift, which, zlib, zstd
+{ stdenv, lib, fetchurl, fetchFromGitHub, fetchpatch, fixDarwinDylibNames
+, autoconf, boost, brotli, cmake, flatbuffers, gflags, glog, gtest, lz4
+, perl , python3, rapidjson, snappy, thrift, utf8proc, which, zlib, zstd
```
Comment (veprbl) [inline pkgs/development/libraries/arrow-cpp/default.nix:3]:
```suggestion
, perl, python3, rapidjson, snappy, thrift, utf8proc, which, zlib, zstd
```

File: pkgs/development/libraries/arrow-cpp/default.nix [hunk 2]
```diff
@@ -1,26 +1,42 @@
...
+
   parquet-testing = fetchFromGitHub {
     owner = "apache";
     repo = "parquet-testing";
     rev = "bcd9ebcf9204a346df47204fe21b85c8d0498816";
     sha256 = "0m16pqzbvxiaradq088q5ai6fwnz9srbap996397znwppvva479b";
   };
 
+  static-utf8proc-patch = fetchpatch {
+    url = "https://patch-diff.githubusercontent.com/raw/apache/arrow/pull/7818.diff";
```
Comment (veprbl) [inline pkgs/development/libraries/arrow-cpp/default.nix:22]: Preferably should reference commit id's, they should have a lesser chance to go stale.

File: pkgs/development/libraries/arrow-cpp/default.nix [hunk 3]
```diff
@@ -32,28 +48,28 @@ in stdenv.mkDerivation rec {
...
-  ] ++ lib.optionals (!enableShared) [
-    # The shared jemalloc lib is unused and breaks in static mode due to missing -fpic.
-    ./jemalloc-disable-shared.patch
   ];
 
+  # Properly exported static targets. Remove at the next version bump.
+  postPatch = ''
+    pushd ..
+    patch -p1 < ${static-utf8proc-patch}
+    patch -p1 < ${static-targets-patch}
```
Comment (veprbl) [inline pkgs/development/libraries/arrow-cpp/default.nix:69]: I believe there is no reason to not use `patches` attribute. The `stripLen` should help to achieve that.

File: pkgs/development/libraries/arrow-cpp/default.nix [hunk 4]
```diff
@@ -32,28 +48,28 @@ in stdenv.mkDerivation rec {
     sha256 = "1xl7z0vwbn5iycg7amka9jd6hxd8nmfk7nahi4p9w2bnw9f0wcrl";
   };
 
+  ARROW_MIMALLOC_URL = fetchurl {
```
Comment (veprbl) [inline pkgs/development/libraries/arrow-cpp/default.nix:51]: So jemalloc is not needed now?

**Review by veprbl — COMMENTED** [review]
I can't review this properly until the week after next, maybe someone else could pick this up. The diff looks good to me, there are few minor comments.

File: pkgs/development/libraries/arrow-cpp/default.nix (hunk 3 above)
Comment (tobim) [reply pkgs/development/libraries/arrow-cpp/default.nix:69]: As far as my knowledge on this goes I'd have to remove the `/cpp` part from `sourceRoot` and cd to the subdirectory in `postPatch`. Maybe there is a better option that I'm missing?

File: pkgs/development/libraries/arrow-cpp/default.nix (hunk 4 above)
Comment (tobim) [reply pkgs/development/libraries/arrow-cpp/default.nix:51]: See https://arrow.apache.org/docs/cpp/api/memory.html?highlight=jemalloc#memory-pools.
jemalloc is still the default, but it is possible to use mimalloc as well. User code gets to choose.

mimalloc support is disabled by default, but enabling it doesn't hurt.

File: pkgs/development/libraries/arrow-cpp/default.nix (hunk 3 above)
Comment (veprbl) [reply pkgs/development/libraries/arrow-cpp/default.nix:69]: There is an option to prepend a string to path. `extraPath`, I think it was

Comment (tobim) [reply pkgs/development/libraries/arrow-cpp/default.nix:69]: I just tried
```
    stripLen = 1;
    extraPrefix = "../";
```
But that fails. `patch` does not seem to like the parent-ref:
```
...
can't find file to patch at input line 72
Perhaps you used the wrong -p or --strip option?
The text leading up to this was:
--------------------------
|--- a/../python/manylinux1/build_arrow.sh
|+++ b/../python/manylinux1/build_arrow.sh
--------------------------
...
```

Comment (veprbl) [reply pkgs/development/libraries/arrow-cpp/default.nix:69]: I believe the "cpp/" prefix needs to be stripped and “python/“ paths need to be filtered (there is an option for that, probably called `exclude`).

Comment (tobim) [reply pkgs/development/libraries/arrow-cpp/default.nix]: Found a way with explicit `includes`, but didn't get to test the static build yet.

Comment (tobim) [reply pkgs/development/libraries/arrow-cpp/default.nix]: Had to make one more adjustment, both `pkgsStatic.arrow-cpp` and `python38Packages.pyarrow` worked for my testcases.

**Comment by ludovicc** [comment]
I'm trying to review this PR on Darwin, but I'm having trouble from some upstream dependencies:

`nix-shell -p nixpkgs-review --run "nixpkgs-review pr 94849"`

[...]

```
$ nix build --no-link --keep-going --option build-use-sandbox relaxed -f /Users/lclaude/.cache/nixpkgs-review/pr-94849-1/build.nix
builder for '/nix/store/cayzbjkbcfpa8sic0vrnjz7v5xajbmb3-arpack-3.7.0.drv' failed with exit code 2; last 10 log lines:
           1 - dnsimp_tst (SEGFAULT)
      2 - bug_1315_single_tst (Child aborted)
        3 - bug_1315_double_tst (Child aborted)
        4 - bug_1323_tst (SEGFAULT)
            5 - bug_58_double_tst (SEGFAULT)
       6 - bug_79_double_complex_tst (SEGFAULT)
       7 - bug_142_tst (SEGFAULT)
     8 - bug_142_gen_tst (SEGFAULT)
  Errors while running CTest
  make: *** [Makefile:114: test] Error 8
cannot build derivation '/nix/store/mc08m835f15a8b8hsq0fwhvhvcqsg9c1-julia-0.7.0.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/bb8yvm3zshizccxnc4d1ymahq998mqi8-julia-1.0.4.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/0xam4h8xzd7qgig0qvxmmw2dccgysz34-julia-1.1.1.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/6jv7ayjf2ywg4g82pv4bw6xppvia678b-julia-1.3.1.drv': 1 dependencies couldn't be built
builder for '/nix/store/dg7x4qi09kw0mfd3ijn0rkmcx9xq6dc2-chronos-maven-deps.drv' failed with exit code 22; last 10 log lines:
    0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
  curl: (22) The requested URL returned error: 501 HTTPS Required
    % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                   Dload  Upload   Total   Spent    Left  Speed
    0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
  curl: (22) The requested URL returned error: 404 Not Found
    % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                   Dload  Upload   Total   Spent    Left  Speed
    0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
  curl: (22) The requested URL returned error: 403 Forbidden
```

**Comment by tobim** [comment]
@ludovicc the final output of `nixpkgs-review` on a macOS 10.14 I tested on is:
```
1 package marked as broken and skipped:
spark

9 packages failed to build:
chronos julia julia_07 julia_11 julia_13 mesos python27Packages.pysvn python37Packages.google_cloud_bigquery python38Packages.google_cloud_bigquery

25 packages built:
arrow-cpp cabal2nix dep2nix git-doc gitFull gitSVN gitAndTools.svn-all-fast-export svn2git nix-prefetch-scripts nix-prefetch-svn nix-update-source perl528Packages.SVNSimple perl530Packages.SVNSimple python37Packages.ibis-framework python37Packages.intake python37Packages.pyarrow python38Packages.intake python38Packages.pyarrow reposurgeon subversion subversionClient subversion_1_10 utf8proc vcstool ydiff
```
The failing packages aren't dependencies of arrow. They rather need to be rebuilt because they are directly or transitively affected by the changes of this PR. My assessments after taking a brief look:
- `google_cloud_bigquery` has some tests failing, probably needs to be updated.
- the julias depend on `arpack` which is broken on darwin. That seems to be an independent issue.
- `mesos` requires `protobuf` but that is missing from the inputs -> seems unrelated.
- Not really sure why `chronos` is being built, probably needs a fix to fetch from `https`.
- `subversion` and related packages are built because of the change to `utf8proc`. `pysvn` exits because it can't find `g++`, I don't think this change is responsible.
`

**Comment by veprbl** [comment]
@tobim The changes look great! Could you, please, reorganize them into two commits with messages according to the convention (e.g. "arrow-cpp: 0.17.1 -> 1.0.0" and "utf8proc: switch to cmake").

**Comment by tobim** [comment]
@veprbl done.

**Review by veprbl — APPROVED** [review]

**Pull request merged**